src/test/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StatusAndTimingTest.java

```diff
@@ -914,6 +916,124 @@ public void parallel() throws Exception {
         }
     }
 
+    @Test
+    @Issue("JENKINS-39203")
+    public void unstableWithWarningAction() throws Exception {
+        WorkflowJob job = j.jenkins.createProject(WorkflowJob.class, "unstable");
+        job.setDefinition(new CpsFlowDefinition(
+                "stage('unstable') {\n" +
+                "  echo('foo')\n" +
+                "  unstable('oops')\n" +
+                "  echo('foo')\n" +
+                "}\n" +
+                "stage('success') {\n" +
+                "  echo('no problem')" +
+                "}\n" +
+                "stage('failure') {\n" +
+                "  unstable('second oops')\n" +
+                "  error('failure')\n" +
+                "}\n", true));
+        WorkflowRun run = j.assertBuildStatus(Result.FAILURE, job.scheduleBuild2(0));
+        FlowExecution exec = run.getExecution();
+        /**
+         * Node dump follows, format:
```

Comment:
Tech debt time! Make a better utility method for these tests that lets the assertions more naturally follow the program text.

Reply:
Done in 74d62cd for the tests I have recently added.